File: codes/cpp/wk02_box_superposition.cpp
// Wk02 - Particle in a box: basis expansion, measurement & dynamics
// Physical Chemistry 2 - Prof. S. Joon Kwon - SPMDL - SKKU
// Build: g++ -O2 -std=c++17 wk02_box_superposition.cpp -o wk02_box_superposition
// hbar = m = L = 1.  c_n -> -8 sqrt(15)/(n^3 pi^3), <E> -> 5.
#include <cmath>
#include <cstdio>
#include <vector>
#include <complex>

int main() {
    const double L = 1.0; const int Nx = 4001, N = 15;
    std::vector<double> x(Nx);
    for (int i = 0; i < Nx; ++i) x[i] = L*i/(Nx-1);
    double h = x[1] - x[0];
    auto trapz = [&](const std::vector<double>& y) {
        double s = 0; for (int i = 0; i < Nx; ++i) s += y[i];
        return h*(s - 0.5*(y[0] + y[Nx-1]));
    };
    auto psi = [&](int n, int i) { return std::sqrt(2.0/L)*std::sin(n*M_PI*x[i]/L); };
    auto En  = [](int n) { return n*n*M_PI*M_PI/2.0; };

    std::vector<double> f(Nx), tmp(Nx);
    for (int i = 0; i < Nx; ++i) f[i] = std::sqrt(30.0)*x[i]*(x[i] - L);
    for (int i = 0; i < Nx; ++i) tmp[i] = f[i]*f[i];
    std::printf("normalization: %.6f\n", trapz(tmp));

    std::vector<double> c(N+1, 0.0);
    for (int n = 1; n <= N; ++n) {
        for (int i = 0; i < Nx; ++i) tmp[i] = psi(n, i)*f[i];
        c[n] = trapz(tmp);
    }
    std::printf(" n   c_n         analytic     |c_n|^2\n");
    for (int n = 1; n <= 7; n += 2)
        std::printf("%2d  %10.6f  %10.6f  %10.6f\n", n, c[n],
                    -8*std::sqrt(15.0)/(n*n*n*M_PI*M_PI*M_PI), c[n]*c[n]);
    double S = 0, Em = 0;
    for (int n = 1; n <= N; ++n) { S += c[n]*c[n]; Em += c[n]*c[n]*En(n); }
    std::printf("Parseval: %.6f,  <E> = %.5f (analytic 5)\n\n", S, Em);

    // superposition (2|1> + |2>)/sqrt(5): density at x = L/4 over one beat
    double c1 = 2/std::sqrt(5.0), c2 = 1/std::sqrt(5.0);
    double Tp = 2*M_PI/(En(2) - En(1));
    std::printf("beat period T = %.4f;  |Psi(L/4,t)|^2:\n  t/T   density\n", Tp);
    for (double frac : {0.0, 0.25, 0.5, 0.75}) {
        double t = frac*Tp, xx = L/4;
        std::complex<double> Psi =
            c1*std::sqrt(2/L)*std::sin(M_PI*xx/L)*std::exp(std::complex<double>(0, -En(1)*t)) +
            c2*std::sqrt(2/L)*std::sin(2*M_PI*xx/L)*std::exp(std::complex<double>(0, -En(2)*t));
        std::printf("%5.2f  %.5f\n", frac, std::norm(Psi));
    }
    std::printf("-> the density sloshes: superpositions move, eigenstates don't.\n");
    return 0;
}

File: codes/julia/wk02_box_superposition.jl
# Wk02 - Particle in a box: basis expansion, measurement & dynamics
# Physical Chemistry 2 - Prof. S. Joon Kwon - SPMDL - SKKU
# hbar = m = L = 1.  c_n = -8 sqrt(15)/(n^3 pi^3) (odd n), <E> = 5.
using Plots, Printf

const L = 1.0
x = range(0, L, length=4001) |> collect
trapz(y) = (x[2]-x[1])*(sum(y) - 0.5*(y[1]+y[end]))
ψ(n) = sqrt(2/L) .* sin.(n*π .* x ./ L)
En(n) = n^2*π^2/2

f = sqrt(30/L^5) .* x .* (x .- L)
@printf("normalization: %.6f\n", trapz(f.^2))

N = 15
c = [trapz(ψ(n) .* f) for n in 1:N]
@printf(" n   c_n         analytic     |c_n|^2\n")
for n in 1:2:7
    @printf("%2d  %10.6f  %10.6f  %10.6f\n", n, c[n], -8sqrt(15)/(n^3*π^3), c[n]^2)
end
@printf("Parseval: %.6f,  <E> = %.5f (analytic 5)\n",
        sum(c.^2), sum(c[n]^2*En(n) for n in 1:N))

p1 = plot(x, f, c=:black, lw=2, label="f(x)", title="Basis expansion")
for Np in (1, 3, 5)
    g = sum(c[n] .* ψ(n) for n in 1:Np)
    plot!(p1, x, g, ls=:dash, label="N=$Np")
end
p2 = bar(1:N, max.(c.^2, 1e-12), yscale=:log10, label=false,
         xlabel="n", title="|c_n|² (measurement prob.)")

c1, c2 = 2/sqrt(5), 1/sqrt(5)
Tp = 2π/(En(2) - En(1))
p3 = plot(title="|Psi(x,t)|² sloshing", xlabel="x")
for frac in (0.0, 0.25, 0.5)
    t = frac*Tp
    Ψ = c1 .* ψ(1) .* exp(-im*En(1)*t) .+ c2 .* ψ(2) .* exp(-im*En(2)*t)
    plot!(p3, x, abs2.(Ψ), lw=1.3, label=@sprintf("t=%.2fT", frac))
end
plot!(p3, x, ψ(1).^2, ls=:dot, c=:black, label="stationary |1>")
display(plot(p1, p2, p3, layout=(1,3), size=(1300,380)))
readline()
